File: apps/notification-service/internal/sender/email.go
package sender

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"

	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
)

const resendEndpoint = "https://api.resend.com/emails"

// Used when EMAIL_FROM is unset. Must name a domain verified in Resend, or
// every send answers 403. A subdomain, so signup complaints never attach to
// the root domain that carries the human mailbox.
const defaultEmailFrom = "KerjaCUS! <noreply@notify.kerjacus.id>"

type EmailSender struct {
	apiKey string
	from   string
	client *http.Client
	// Overridden only by tests; production always talks to Resend.
	baseURL string
}

// NewEmailSender builds the Resend client. from is the RFC 5322 From header and
// its domain must be the one verified with Resend; an unverified domain is
// rejected upstream, not here.
func NewEmailSender(apiKey, from string) *EmailSender {
	if from == "" {
		from = defaultEmailFrom
	}
	return &EmailSender{
		apiKey:  apiKey,
		from:    from,
		baseURL: resendEndpoint,
		client: &http.Client{
			Timeout:   10 * time.Second,
			Transport: otelhttp.NewTransport(http.DefaultTransport),
		},
	}
}

type SendEmailInput struct {
	To      string
	Subject string
	HTML    string
}

type resendRequest struct {
	From    string   `json:"from"`
	To      []string `json:"to"`
	Subject string   `json:"subject"`
	HTML    string   `json:"html"`
}

// ErrNotConfigured reports that no Resend key is set, so nothing can be sent.
//
// It used to be a warning and a nil return: an unconfigured deployment
// reported every send as a success it never attempted, which is how a password
// reset or a verification mail went missing with nothing to retry, nothing in
// the dead letter queue and nothing on screen - recovery answers identically
// whether the address exists. An error stops at the consumer, which naks the
// event and finally parks it where an operator can see it.
var ErrNotConfigured = errors.New("RESEND_API_KEY is not configured")

func (s *EmailSender) Send(ctx context.Context, in SendEmailInput) error {
	if s.apiKey == "" {
		return ErrNotConfigured
	}

	body := resendRequest{
		From:    s.from,
		To:      []string{in.To},
		Subject: in.Subject,
		HTML:    in.HTML,
	}

	payload, err := json.Marshal(body)
	if err != nil {
		return fmt.Errorf("marshal email request: %w", err)
	}

	url := s.baseURL
	if url == "" {
		url = resendEndpoint
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(payload))
	if err != nil {
		return fmt.Errorf("create email request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+s.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		return fmt.Errorf("send email: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("resend API error (status %d): %s", resp.StatusCode, string(respBody))
	}

	slog.Info("email sent", "to", in.To, "subject", in.Subject)
	return nil
}

File: apps/auth-service/src/lib/email.ts
/**
 * Transactional mail for the auth flows: verification and password recovery.
 *
 * Both env reads happen per call, not at module load. A module constant froze
 * the value at import, which made the configured and unconfigured paths two
 * different module instances and left the choice to whether whoever ran the
 * process happened to have a .env -- it covered locally and not in CI.
 */

/**
 * Used when EMAIL_FROM is unset. Must name a domain verified in Resend, or
 * every send answers 403 and the failure looks like silence: password recovery
 * replies identically whether the address exists, so nothing on screen can
 * tell a rejected send from a delivered one.
 *
 * A subdomain, because signup sends to addresses nobody confirmed wants mail
 * and those complaints attach to whichever domain signs the DKIM. The root
 * carries the human mailbox. This read RESEND_FROM before, a name set nowhere
 * in the repo.
 */
const DEFAULT_EMAIL_FROM = 'KerjaCUS! <noreply@notify.kerjacus.id>'

export type SendEmailParams = {
  to: string
  subject: string
  html: string
  text?: string
}

/**
 * `b***@example.com`. Enough to tell which send is which in a log, not enough
 * to harvest the address list out of one.
 */
function redactAddress(address: string): string {
  const at = address.lastIndexOf('@')
  if (at < 1) return '***'
  return `${address[0]}***${address.slice(at)}`
}

export async function sendEmail(params: SendEmailParams): Promise<void> {
  const apiKey = process.env.RESEND_API_KEY
  if (!apiKey) {
    /*
     * The envelope only.
     *
     * This used to log `params`, and params carries the body: for a password
     * reset and for email verification that body is a single-use link with the
     * token in it. Production runs without RESEND_API_KEY, so every reset
     * request wrote a working account-takeover link into a log stream that far
     * more people can read than can read the mailbox it was meant for.
     */
    console.log('[email] RESEND_API_KEY missing, message dropped:', {
      to: redactAddress(params.to),
      subject: params.subject,
    })
    return
  }
  const res = await fetch('https://api.resend.com/emails', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({ from: process.env.EMAIL_FROM || DEFAULT_EMAIL_FROM, ...params }),
  })
  if (!res.ok) {
    const err = await res.text().catch(() => '')
    throw new Error(`Resend send failed: ${res.status} ${err}`)
  }
}

export function buildVerificationEmail(name: string, verifyUrl: string) {
  return {
    subject: 'Verifikasi email KerjaCUS Anda',
    html: `<p>Hi ${name},</p><p>Klik link berikut untuk verifikasi email:</p><p><a href="${verifyUrl}">${verifyUrl}</a></p><p>Link berlaku 24 jam.</p>`,
    text: `Hi ${name}, verifikasi email Anda di: ${verifyUrl}`,
  }
}
